Support Min P Sampler

esmeetu opened the pull request:
### Reference: 
https://github.com/oobabooga/text-generation-webui/pull/4449
https://github.com/ggerganov/llama.cpp/pull/3841

It seems sample better result than normal top_k/top_p method. And i tested min_p = 0.05 when inferencing code LLM, it gave me a better explanation.

esmeetu committed:
feat: support min_p sampler

esmeetu committed:
add doc

esmeetu committed:
format

zhuohan123 commented on vllm/model_executor/layers/sampler.py:
```diff
@@ -313,6 +322,28 @@ def _apply_top_p_top_k(
+    scaled_min_p = min_p.unsqueeze(dim=1) * top_probs
+    tokens_to_remove = probs < scaled_min_p
+
+    sorted_indices = torch.argsort(logits, descending=True, dim=-1)
+    sorted_indices_to_remove = torch.gather(tokens_to_remove,
+                                            dim=-1,
+                                            index=sorted_indices)
```
Seems like the sort here is unnecessary? I believe we only need to mask out the positions where `probs < scaled_min_p`?

zhuohan123 commented on vllm/model_executor/layers/sampler.py:
```diff
@@ -313,6 +322,28 @@ def _apply_top_p_top_k(
+def _apply_min_p(
+    logits: torch.Tensor,
+    min_ps: List[float],
+) -> torch.Tensor:
```
Add a reference to the original implementation here?

esmeetu committed:
fix comments

esmeetu commented:
@zhuohan123 Thanks for you reviews! In terms of first problem, i removed the `min_tokens_to_keep ` params which need sorting to take from original logits. Why i removed that params is because i want to keep this implement simpler, and align with top_p/top_k.

zhuohan123 reviewed (approved):
LGTM! Thank you for your great contribution!
